### PR TITLE
DoE: Fix bug if fixed_experiments contain columns that are not in domain

### PR DESCRIPTION
**Pull request by dlinzner-bcs**
from here: https://github.com/experimental-design/bofire/issues/320

### CODE REVIEW

**Comment by jduerholt** [comment]
Just one question from my side: why is this checking necessary? When you work with the fixed and partially fixed dataframes, you could just use the input columns which are part of the domain by always stating `fixed_experiments[domain.inputs.get_keys()]`, or not? This would be bit more flexible, what do you think?

**Review by jduerholt — APPROVED** [review]

**Comment by jduerholt** [comment]
One question: what happens if the the intersection is empty?

**Comment by dlinzner-bcs** [comment]
> One question: what happens if the the intersection is empty?

The logic is now: 1. Check if all needed columns are present 2. Project dataframe on columns in `domain.inputs.get_keys()`
if 1. fails, you get an informative value error.

**Comment by jduerholt** [comment]
Somehow tests are failing again ...

File: tests/bofire/strategies/doe/test_design.py [hunk 1]
```diff
@@ -550,3 +550,122 @@ def test_partially_fixed_experiments():
...
+    assert doe.shape == (4, 4)
+    assert np.allclose(doe.iloc[[0, 1]]["x1"], fixed_experiments["x1"])
+    assert np.allclose(doe.iloc[[0, 1]]["x2"], fixed_experiments["x2"])
+    assert np.allclose(doe.iloc[[0, 1]]["a1"], fixed_experiments["a1"])
+    assert np.allclose(doe.iloc[[0, 1]]["a2"], fixed_experiments["a2"])
+    assert np.allclose(
+        doe.iloc[[2, 3]]["x1"], partially_fixed_experiments["x1"].astype(float)
+    )
+
+    with pytest.raises(ValueError):
```
Comment (jduerholt) [inline tests/bofire/strategies/doe/test_design.py:625]: Ideally, we would also catch here the exact error message to be sure that we are catching the correct stuff.

File: bofire/strategies/doe/design.py [hunk 2]
```diff
@@ -402,6 +402,33 @@ def find_local_max_ipopt(
...
         raise e
 
+    # determine number of experiments (only relevant if n_experiments is not provided by the user)
+    n_experiments = get_n_experiments(
+        domain=domain, model_type=model_type, n_experiments=n_experiments
+    )
+
+    if partially_fixed_experiments is not None:
+        # check if partially fixed experiments are valid
+        check_partially_fixed_experiments(
```
Comment (jduerholt) [inline bofire/strategies/doe/design.py:412]: Do we want to have a test just for these added methods?

File: bofire/strategies/doe/design.py [hunk 3]
```diff
@@ -402,6 +402,33 @@ def find_local_max_ipopt(
...
+            domain, n_experiments, partially_fixed_experiments
+        )
+        # no columns from partially fixed experiments which are not in the domain
+        partially_fixed_experiments = partially_fixed_experiments[
```
Comment (jduerholt) [inline bofire/strategies/doe/design.py:416]: One could also set `partially_fixed_experiments` as the return value of the check method. We do the same in domain for validate_experiments and validate_candidates.

File: bofire/strategies/doe/design.py [hunk 4]
```diff
@@ -583,53 +605,76 @@ def partially_fix_experiment(
...
-    n_fixed_experiments, D = np.array(fixed_experiments).shape
+    n_fixed_experiments = len(fixed_experiments.index)
 
     if n_fixed_experiments >= n_experiments:
         raise ValueError(
             "For starting the optimization the total number of experiments must be larger that the number of fixed experiments."
         )
 
-    if D != len(domain.inputs):
+    input_set = set(domain.inputs.get_keys())
```
Comment (jduerholt) [inline bofire/strategies/doe/design.py:624]: You could also somehow solve this by `domain.validate_experiments(only_inputs=True)`

**Review by jduerholt — APPROVED** [review]
Looks good to me, I only let a few minor comments

File: bofire/strategies/doe/design.py (hunk 3 above)
Comment (dlinzner-bcs) [reply bofire/strategies/doe/design.py:416]: this does not seem to project out columns not in domain.inputs.

File: bofire/strategies/doe/design.py (hunk 4 above)
Comment (dlinzner-bcs) [reply bofire/strategies/doe/design.py]: Did this. Required to change the validate_candidates function slightly (add additional flags) as `partially_fixed_experiments` will contain `None`.

Comment (dlinzner-bcs) [reply bofire/strategies/doe/design.py]: Reverted these changes. Only use `validate_candidates` for `fixed_experiments` now - where the function can stay as it is.

**Pull request merged**